### ISSUE TITLE
Question: How did you manage to catch the exceptions from the auto closures?

### ISSUE THREAD

**Issue by silkentrance** [Quick/Nimble#108]
Aren't these executed out of context of the @try/@catch ...?

I am asking because I tried implementing something very similar in order to be able to extend NSException with a try...catch..finally construct that would allow me to use exceptions in swift.
However, all my tests failed since the closures I passed in would be executed lazily while my ObjectiveC based category would never catch the exceptions being thrown.
Please note that I did not use auto closures...

TIA for enlightening me...

**Comment by jeffh** [comment]
Hey @silkentrance,

I'm not sure I full understand but this is roughly what Nimble does:

- It creates an implicit closure using `@autoclosure` to create a block.
- The `raiseException` matcher calls this closure inside a [try-catch call](https://github.com/Quick/Nimble/blob/master/Nimble/Matchers/RaisesException.swift#L15-22).

Since Swift doesn't support catching exceptions, they're implemented in objc.

**Comment by silkentrance** [comment]
Yes, this is exactly what I implemented, except for me not using the auto closures but rather

```
NSException.try({
     NSException(...).raise()
}, catch: {
     // ...
})
```

And in the ObjC part then something that is very similar to your implementation.

Now, when using it like this, the exception will not be caught by the ObjC part, thus never entering the catch block. It seems as if `@autoclosure` circumvents this.

**Comment by jeffh** [comment]
It seems like it should work otherwise. Without more thorough code examples I can't really say much else. I'm assuming you're not assuming that Swift code will raise exceptions, since swift assertions simply abort the program execution.

**Comment by silkentrance** [comment]
It turned out, that I did use

```
NSException.raise(NSInvalidArgumentException, format: nil, ...);
```

inside of the block I passed to NSException.try. Since non of the arguments is optional, I shouldn't have passed in nil. Turns out that the compiler will not report that mistake but rather report some entirely different error.

As for the problem with the exceptions not being caught, it simply went away after setting up the project from scratch and recompiling everything.

Problem solved.
